src/lib.rs

```diff
@@ -91,72 +91,48 @@ pub struct Gc<T> {
 impl<T> Gc<T> {
     /// Constructs a new `Gc<T>`.
     pub fn new(v: T) -> Self {
-        let objptr = Self::alloc_blank(Layout::new::<T>());
-        let mut gc = unsafe {
-            objptr.copy_from_nonoverlapping(&v, 1);
-            Gc::from_raw(objptr)
-        };
-
-        forget(v);
-
-        let to: &dyn Trace = &gc;
-        let vptr = unsafe { transmute::<*const dyn Trace, (usize, usize)>(to).1 };
-        gc.set_header(GcHeader::new(vptr));
-
-        gc
-    }
-
-    /// Create a `Gc` from a raw pointer previously created by `alloc_blank` or
-    /// `into_raw`.
-    pub unsafe fn from_raw(objptr: *const T) -> Self {
+        let boxed = GcBox::new(v);
         Gc {
-            objptr: objptr as *mut T,
+            objptr: boxed as *mut T,
         }
     }
 
     /// Get a raw pointer to the underlying value `T`.
     pub fn as_ptr(&self) -> *const T {
         self.objptr
     }
+}
 
-    /// Returns a reference to the header of the `Gc<T>`. The header stores
-    /// meta-data about the value used by the collector. For example, the
-    /// mark-bit used to denote the object's reachability, is stored in
-    /// the header.
-    fn header(&self) -> GcHeader {
-        let raw_bytes = unsafe {
-            let headerptr = (self.objptr as *const usize).sub(1) as *mut [u8; 8];
-            std::ptr::read(headerptr)
-        };
-        GcHeader::unpack(&raw_bytes).unwrap()
-    }
+/// A `GcBox` is a 0-cost wrapper which allows a single `Drop` implementation
+/// while also permitting multiple, copyable `Gc` references. The `drop` method
```

Comment:
Is it just me or is there not actually a `drop` method? I guess you mean something like "Note that `GcBox` intentionally does not define its own `drop` method, since dropping it should do nothing: only when the GC later determines that nothing points to the underlying object will that object's `drop` method be called."?

Reply:
There is a `Drop` method, but as you noticed below I included it as part of the wrong commit ;). I'll fix this as part of squashing.